src/sdam/monitor.rs

```diff
@@ -184,7 +200,46 @@ impl HeartbeatMonitor {
         let result = match self.connection {
             Some(ref mut conn) => {
                 let command = is_master_command(self.client_options.server_api.as_ref());
-                run_is_master(command, conn).await
+                emit_event!(
+                    self,
+                    ServerHeartbeatStartedEvent {
+                        server_address: self.address.clone(),
+                    },
+                    handle_server_heartbeat_started_event
+                );
+
+                let start_time = Instant::now();
+                let response_result = conn.send_command(command, None).await;
+                let end_time = Instant::now();
+
+                let round_trip_time = end_time.duration_since(start_time);
+
+                match response_result {
+                    Ok(raw_response) => match raw_response.as_is_master_response(round_trip_time) {
+                        Ok(is_master_reply) => {
+                            emit_event!(
+                                self,
+                                ServerHeartbeatSucceededEvent {
+                                    duration: round_trip_time,
+                                    reply: raw_response.body::<Document>().unwrap_or_else(
+                                        |err| doc! { "deserialization error": err.to_string() }
+                                    ),
+                                    server_address: self.address.clone(),
+                                },
+                                handle_server_heartbeat_succeeded_event
+                            );
+                            Ok(is_master_reply)
+                        }
+                        Err(err) => {
+                            self.emit_heartbeat_failed_event(err.clone(), round_trip_time);
+                            Err(err)
+                        }
+                    },
+                    Err(err) => {
+                        self.emit_heartbeat_failed_event(err.clone(), round_trip_time);
+                        Err(err)
+                    }
+                }
```

Comment:
I believe we need to also emit events in the case that the heartbeat is performed as part of creating the monitoring connection (i.e. the `None` branch here), so we'll need to lift that logic out of the `match` entirely and possibly add a new one. This may alleviate that immutable / mutable borrow issues actually, since the mutable borrow of `conn` will no longer be in scope.

Reply:
So this ended up making things a bit more complicated because of how establishing a new connection and `Handshaker::handshake` work. A few issues that arose:

1. Creating a new connection can possibly return an error. In this case, we don't want to emit any SDAM events (including a `ServerHeartbeatStartedEvent`) because we won't be sending anything to the server. This means we can't emit the `ServerHeartbeatStartedEvent` prior to the `match` expression on `self.connection` because we're not sure we'll emit a corresponding success/failure event. Alternatively, we could emit this event within each branch of the `match` expression, but ideally we'd avoid duplicating these events and we would still encounter the ownership issue that requires the macro. Emitting the event after the `match` expression would technically emit the right events in the right order, but that would be out of compliance with the spec, which states that the `ServerHeartbeatStartedEvent` should be emitted prior to sending the command to the server.
2. `Handshaker::handshake` can possibly return an error when setting speculative auth info. This is a similar issue to the one presented by the incorrect command name error we removed from `run_is_master`, but this one cannot be removed. Similar to above, we need to ensure we make it past this error prior to emitting a `ServerHeartbeatStartedEvent`.
3. `Handshaker::handshake` uses the `IsMasterReply` parsed from the `RawCommandResponse` for updating the connection's stream description and moving the speculative authentication information from the `IsMasterCommandResponse` to a `FirstRound`. This means we'd need to move this logic out of the method to avoid returning both an `IsMasterReply` and `RawCommandResponse` or converting to an `IsMasterReply` twice.

Given these issues, we'd need to inline a pretty large chunk of `Handshaker::handshake` everywhere it's currently being called in order to catch the right errors and extract the necessary information. I don't think this is a great idea because this will require making duplicate changes in two places any time we need to update our handshake logic, which can be quite error-prone.

To that end, I decided to reintroduce `run_is_master` and contain all of the SDAM event logic in that function. This required passing in the `WeakTopology` to ensure that events are only emitted while the topology is still alive. The `Some` branch of the `match` on `self.connection` and `Handshaker::handshake` both call this method; the latter now takes optional arguments for the topology and handler which are propagated as necessary from `perform_is_master`. This played more nicely with the borrow checker so I was able to implement `emit_event` as a function rather than a macro.

Reply:
All good points. I think your solution of reintroducing `run_is_master` with SDAM monitoring included looks good. Sorry for leading you down all these rabbit holes! I think one big lesson learned from all this (which will come in handy for Swift) is that the SDAM implementation and SDAM monitoring implementation should happen at the same time, rather than one and then the other much later, since it can be hard to figure out where this stuff needs to live after the fact.